topdown/parse_units.go

```diff
@@ -0,0 +1,114 @@
+// Copyright 2016 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+package topdown
+
+import (
+	"fmt"
+	"math/big"
+	"strings"
+
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/topdown/builtins"
+)
+
+// Binary Si unit constants are borrowed from topdown/parse_bytes
+const milli float64 = 0.001
+const (
+	k uint64 = 1000
+	m        = k * 1000
+	g        = m * 1000
+	t        = g * 1000
+	p        = t * 1000
+	e        = p * 1000
+)
+
+func parseUnitsError(msg string) error {
+	return fmt.Errorf("%s error: %s", ast.UnitsParseUnits.Name, msg)
+}
+
+func errUnitNotRecognized(unit string) error {
+	return parseUnitsError(fmt.Sprintf("Unit %s not recognized", unit))
+}
+
+var (
+	errNoAmount       = parseUnitsError("no amount provided")
+	errNumConv        = parseUnitsError("could not parse amount to a number")
+	errIncludesSpaces = parseUnitsError("spaces not allowed in resource strings")
+)
+
+// Accepts both normal SI and binary SI units.
+func builtinUnits(bctx BuiltinContext, operands []*ast.Term, iter func(*ast.Term) error) error {
+	var x big.Float
+
+	raw, err := builtins.StringOperand(operands[0].Value, 1)
+	if err != nil {
+		return err
+	}
+
+	s := string(raw)
+	s = strings.Replace(s, "\"", "", -1)
```

Comment:
To be honest, I just inlined the body of `topdown/parse_bytes`'s  `formatString` function there, so that's where the quotes replacement logic came from. I'll take a second look to see if it seems to be really necessary.

Reply:
So, after checking: There are actual test cases with instances like `units.parse("\"100TI\"") == 109951162777600` under `test/cases/testdata/units/test-parse-units.yaml` (and in `test-parse-bytes.yaml` as well).

I *believe* these are around because of the possibility of resource strings getting quoted and escaped by accident in incoming JSON blobs.

Reply:
I thought the same thing about `formatString` originally! The symmetry between the two modules would've been great.

However, `formatString` strips quotes *and* lowercases the entire string. So, when I realized I couldn't use the function as-is, I chose to split its logic in half: the quote stripping logic remains at where the original `formatString` call site was, and the lowercasing logic lives a few lines further down to ensure that we leave the first character of the unit alone. (so that we can distinguish 'm' vs 'M')

Would you recommend re-condensing things back into a function? (maybe something like `formatUnitsString`?) Or would something like additional comments (documenting the source of the quote stripping) be the better resolution here?

Reply:
OK thanks I wasn't aware of those details 😃 

> Or would something like additional comments (documenting the source of the quote stripping) be the better resolution

Yeah let's do that... add a comment saying that it's for parity with `units.parse_bytes`.